### ISSUE TITLE
Can Not Stop Running Sketch

### ISSUE THREAD

**Issue by KaroAntonio** [processing/p5.js-editor#37]
Cannot Stop a Sketch to save or update the index or sketch files, the browser window remains open and the sketch will not terminate. The only way to continue working on the sketch is to re-open it. 

![no stop](https://cloud.githubusercontent.com/assets/5997028/6653146/03102f2a-ca4c-11e4-8c4c-fefadc98d361.png)

**Comment by mactkg** [comment]
@KaroAntonio Can you paste sources here?

**Comment by KaroAntonio** [comment]
Sources? Sorry I'm not quite familiar with what exactly sure what those would be but here's some context info!

OS: OSC 10.9.5
Hardware: MacBook Air
Browser: Safari

**Comment by sepans** [comment]
Isn't this the same infinite loop problem, @antiboredom?

if so, I noticed that codepen injects some checkpoint code to the loop to break it.
<pre>
for (var i = 0; i < data.length; i++) {
    //CodePen injexted check
    if (window.CP.shouldStopExecution(1)) {
        break;
    }
    /* ... user code ... */
 }
//CodePen injected check
window.CP.exitedLoop(1)
</pre>

**Comment by antiboredom** [comment]
Could be! How do we feel about injecting code this way. I'm sort of hesitant (but open to it). Are there any other solutions? Also what does the processing IDE in these situations?

**Comment by shiffman** [comment]
Interesting, the Processing IDE does not inject any code with a couple exceptions.  Tweak mode rewrites your code behind the scenes to allow for real-time manipulation of variables.  And things are moved around to deal with make the sketch a wacky-looking java class (`size()` is moved behind the scenes to a `settings()` function etc.)

I would say, injecting code into each and every loop to protect against an infinite loop is going too far for the desktop IDE, although I see why that might be necessary in the context of a web-based editor like codepen. @therewasaguy, something to think about for the web IDE in terms of protecting the user from  crashing the browser.

In the case of a frozen sketch (like with an infinite loop) Processing's stop function is able to kill the sketch.  It would be nice to have that behavior here so that the IDE does not crash, losing any recent changes by the user.

**Comment by sepans** [comment]
The accidental infinite loops are not very uncommon so maybe this is a higher priority bug (and we can move the discussion to a new issue.)

I just wonder why even the editor crashes when there is a infinite loop in the code that runs only in the browser. Are both editor and browser window running in the same process @antiboredom? If so maybe it is possible to have nw fork the process and it should fix the problem, similar to Processing IDE (pushing stop button).

Totally agree @shiffman @antiboredom that injecting code should be the last resort. But it is not a very terrible thing because the injected code is not visible to users unless they go to browser developer-tools and view sketch.js source code from there (which I cannot think of any use case for doing that except for debugging the code using developer-tools)

**Comment by antiboredom** [comment]
Yeah - the output windows are running in the same process - and as @sepans says forking could definitely fix it. I had tried to do this on an earlier iteration of the project but ran into some difficulties... it's worth another look.
